aten/src/ATen/xpu/CachingHostAllocator.cpp

```diff
@@ -0,0 +1,77 @@
+#include <ATen/xpu/CachingHostAllocator.h>
+
+namespace at::xpu {
+namespace {
+
+constexpr size_t kHostAlignment = 512;
+
+using Block = HostBlock<XPUStream>;
+using AllocatorImplInterface =
+    CachingHostAllocatorImplInterface<XPUStream, XPUEvent>;
+
+struct XPUHostAllocatorImpl : public AllocatorImplInterface {
+  /* These following functions are runtime-related. */
+  void allocate_host_memory(size_t size, void** ptr) override {
+    *ptr = sycl::aligned_alloc_host(
+        kHostAlignment, size, c10::xpu::get_device_context());
+  }
+
+  void free_block(Block* block) override {
+    sycl::free(block->ptr_, c10::xpu::get_device_context());
+  }
+
+  void record_stream(
+      c10::optional<std::vector<XPUEvent>>& events,
+      XPUStream stream) override {
+    XPUEvent event;
+    event.record(stream);
+    events->push_back(std::move(event));
+  }
+
+  bool query_event(XPUEvent& event) override {
+    return event.query();
+  }
+};
+
+} // anonymous namespace
+
+void raw_local_deleter(void* ptr);
+
+struct XPUHostAllocator final
+    : public HostAllocatorInterface<XPUHostAllocatorImpl> {
+  at::DataPtr allocate(size_t size) override {
+    auto ptr_and_ctx = impl_->allocate(size);
+    return {
+        ptr_and_ctx.first,
+        ptr_and_ctx.second,
+        &raw_local_deleter,
+        at::DeviceType::CPU};
+  }
```

Comment:
There is no difference between CUDA and XPU. Is it possible to move the function to the parent class?

Reply:
`raw_local_deleter` is a deleter, which is necessary to define by an instance `caching_host_allocator`. So we need a forward declaration. If we move this function to the parent class,  we have to define `raw_local_deleter` under namespace `at` for each backend. There will be duplicated definitions for `raw_local_deleter` under the same namespace `at` even though the current PyTorch is the exclusive backend. It is not elegant.